File: p238.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        int after[n];
        after[n-1] = nums[n-1];
        for(int j = nums.size()-2; j >= 0; --j) {
            after[j] = nums[j]*after[j+1];
        }
        vector<int> output;
        int accumulate = nums[0];
        output.push_back(after[1]);
        for(int k = 1; k < nums.size()-1; ++k) {
            output.push_back(accumulate * after[k+1]);
            accumulate *= nums[k];
        }
        output.push_back(accumulate);
        return output;
    }
};

File: p238.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        left = nums[:]
        right = nums[:]

        for i in range(1, len(nums)):
            left[i] *= left[i-1]
            right[-i-1] *= right[-i]

        left = [1] + left[:] + [1]
        right = [1] + right[:] + [1]

        ans = []

        for i in range(len(nums)):
            ans.append(left[i] * right[i+2])

        return ans
